ykcapi/src/sginterp.rs

```diff
@@ -0,0 +1,142 @@
+use llvm_sys::bit_reader::LLVMParseBitcodeInContext2;
+use llvm_sys::core::*;
+use llvm_sys::prelude::{LLVMModuleRef, LLVMValueRef};
+use llvm_sys::{LLVMOpcode};
+use std::collections::HashMap;
+use std::ffi::CStr;
+use std::mem::MaybeUninit;
+
+use crate::llvmapihelper::{self};
+
+/// Stopgap interpreter values.
+#[derive(Debug)]
+pub enum SGValue {
+    U32(u32),
+    U64(u64),
+}
+
+/// A frame holding live variables.
+struct Frame {
+    vars: HashMap<LLVMValueRef, SGValue>,
+}
+
+impl Frame {
+    fn new() -> Frame {
+        Frame {
+            vars: HashMap::new(),
+        }
+    }
+
+    /// Get the value of the variable `key` in this frame.
+    fn get(&self, key: &LLVMValueRef) -> Option<&SGValue> {
+        self.vars.get(key)
+    }
+
+    /// Insert new variable into this frame.
+    fn add(&mut self, key: LLVMValueRef, val: SGValue) {
+        self.vars.insert(key, val);
+    }
+}
+
+/// The stopgap interpreter. Used during guard failures to get back to the control point by
+/// interpreting LLVM IR.
+pub struct SGInterp {
+    /// LLVM IR module we are interpreting.
+    module: LLVMModuleRef,
+    /// Current frames.
+    frames: Vec<Frame>,
+    /// Current instruction being interpreted.
+    pc: LLVMValueRef,
+}
+
+impl SGInterp {
+    /// Create a new stopgap interpreter and initialise it to start interpretation at the location
+    /// given by a basic block index, instruction index, and function name.
+    /// FIXME: Support initalisation of multiple frames.
+    pub unsafe fn new(bbidx: u32, instridx: u32, fname: &CStr) -> SGInterp {
+        // Get AOT module IR and parse it.
+        let (addr, size) = ykutil::obj::llvmbc_section();
+        let membuf = LLVMCreateMemoryBufferWithMemoryRange(
+            addr as *const i8,
+            size,
+            "".as_ptr() as *const i8,
+            0,
+        );
+        let context = LLVMContextCreate();
+        let mut module: MaybeUninit<LLVMModuleRef> = MaybeUninit::uninit();
+        let module = {
+            LLVMParseBitcodeInContext2(context, membuf, module.as_mut_ptr());
+            module.assume_init()
+        };
+        // Create and initialise stop gap interpreter.
+        let func = LLVMGetNamedFunction(module, fname.as_ptr());
+        let bb = llvmapihelper::get_basic_block(func, bbidx);
+        let instr = llvmapihelper::get_instruction(bb, instridx);
+        SGInterp {
+            module,
+            frames: vec![Frame::new()],
+            pc: instr,
+        }
+    }
+
+    /// Add a live variable and its value to the current frame.
+    pub unsafe fn init_live(&mut self, bbidx: u32, instridx: u32, fname: &CStr, value: SGValue) {
```

Comment:
I used to call it `add_var`, but then I was worried the function will be confusing later, since this one is only to add variables to he frame during initialisation of the stopgap interpreter. Maybe `add_live` is a good compromise.